### PR TITLE
Feature/kinesis spark streaming source

### PR DESCRIPTION
**Pull request by abhinavbansal246**
Adding Kinesis spark streaming source for hydrator

### CODE REVIEW

File: spark-plugins/widgets/KinesisSource-streamingsource.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        {
+          "widget-type": "textbox",
+          "label": "AWS access secret",
+          "name": "awsAccessSecret"
+        }
+      ]
+    }
+  ],
+  "outputs": []
+}
```
Comment (abhinavbansal246) [inline spark-plugins/widgets/KinesisSource-streamingsource.json]: I will put eof newline

File: core-plugins/pom.xml [hunk 2]
```diff
@@ -164,10 +164,6 @@
           <groupId>com.fasterxml.jackson.core</groupId>
           <artifactId>jackson-core</artifactId>
         </exclusion>
-        <exclusion>
```
Comment (rohitsinha54) [inline core-plugins/pom.xml:167]: Why remove this exclusion ?

File: spark-plugins/docs/KinesisSource-streamingsource.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+# Amazon Kinesis Stream spark streaming source
```
Comment (rohitsinha54) [inline spark-plugins/docs/KinesisSource-streamingsource.md]: Weird capitalization

File: spark-plugins/docs/KinesisSource-streamingsource.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Description
+-----------
+Spark streaming source that reads from AWS kinesis streams.
+
+Use Case
+--------
+This source is used when you want to read data from a kinesis stream in real-time. For example, you
+may want to read data from a kinesis stream write it to a cdap dataset.
+
```
Comment (rohitsinha54) [inline spark-plugins/docs/KinesisSource-streamingsource.md]: Sometimes you use one line for separation sometimes two. Better to keep one format.

File: spark-plugins/docs/KinesisSource-streamingsource.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+Properties
+----------
+**appName:** The name of the kinesis application. The application name that is used to checkpoint the Kinesis sequence
+numbers in DynamoDB table
+
+**streamName:** The name of the Kinesis stream to the get the data from. The stream should be active.
+
+**duration:** The interval in milliseconds at which the Kinesis Client Library saves its position in the stream.
+
+**endpointUrl:** Valid Kinesis endpoints URL eg. kinesis.us-east-1.amazonaws.com.
```
Comment (rohitsinha54) [inline spark-plugins/docs/KinesisSource-streamingsource.md]: better to remove . after com to avoid confusion

File: spark-plugins/docs/KinesisSource-streamingsource.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+**accessID:** The access Id provided by AWS required to access the kinesis streams.
```
Comment (rohitsinha54) [inline spark-plugins/docs/KinesisSource-streamingsource.md]: Will these be securely stored with key management ? Better to mention that here in a line for assurance to user.

File: spark-plugins/pom.xml [hunk 7]
```diff
@@ -22,13 +22,11 @@
     <groupId>co.cask.hydrator</groupId>
     <version>1.5.0-SNAPSHOT</version>
   </parent>
-
   <properties>
     <app.parents>system:cdap-data-pipeline[4.0.0-SNAPSHOT,4.1.0-SNAPSHOT),system:cdap-data-streams[4.0.0-SNAPSHOT,4.1.0-SNAPSHOT)</app.parents>
-    <spark.version>1.3.1</spark.version>
+    <spark.version>1.6.2</spark.version>
```
Comment (rohitsinha54) [inline spark-plugins/pom.xml:27]: Why the bump ?

File: spark-plugins/pom.xml [hunk 8]
```diff
@@ -72,6 +70,23 @@
...
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/org.apache.spark/spark-streaming_2.10 -->
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-streaming_2.10</artifactId>
+      <version>1.6.2</version>
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient -->
+    <dependency>
+      <groupId>org.apache.httpcomponents</groupId>
```
Comment (rohitsinha54) [inline spark-plugins/pom.xml:78]: why is this needed ?

File: spark-plugins/pom.xml [hunk 9]
```diff
@@ -72,6 +70,23 @@
...
     <dependency>
       <groupId>org.apache.spark</groupId>
+      <artifactId>spark-streaming-kinesis-asl_2.10</artifactId>
+      <version>1.6.2</version>
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/org.apache.spark/spark-streaming_2.10 -->
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-streaming_2.10</artifactId>
+      <version>1.6.2</version>
```
Comment (rohitsinha54) [inline spark-plugins/pom.xml]: Can we exclude things from this which we don't need ?

File: spark-plugins/pom.xml [hunk 10]
```diff
@@ -177,7 +192,14 @@
...
           <instructions>
             <_exportcontents>
-              co.cask.hydrator.plugin.spark.*;org.apache.spark.streaming.kafka.*;org.apache.spark.streaming.scheduler.*;kafka.serializer.*;org.apache.spark.streaming.twitter.*;twitter4j.*;
+              co.cask.hydrator.plugin.spark.*;
+              org.apache.spark.streaming.kafka.*;
+              com.amazonaws.services.kinesis.*;
+              org.apache.spark.streaming.kinesis.*;
+              org.apache.spark.streaming.scheduler.*;
+              kafka.serializer.*;org.apache.spark.streaming.twitter.*;
+              twitter4j.*;com.amazonaws.*;
```
Comment (rohitsinha54) [inline spark-plugins/pom.xml]: if breaking in independent lines break all of them.

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    recordBuilder.set(BODY, new String(data));
+    return recordBuilder.build();
+  }
+
+  /**
+   * config file for Kinesis stream sink
+   */
+  public static class KinesisStreamConfig extends ReferencePluginConfig implements Serializable {
+
+    @Name("appName")
```
Comment (rohitsinha54) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java:190]: this and all the fields below should be final.

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    return kinesisStream.map(new Function<byte[], StructuredRecord>() {
+                               public StructuredRecord call(byte[] data) {
+                                 return recordTransform(data);
+                               }
+                             }
+    );
+  }
+
+  private StructuredRecord recordTransform(byte[] data) {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(SCHEMA);
```
Comment (rohitsinha54) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java]: Why is the SCHEMA just a single string. I remember that we discussed that you should be doing something similar to what KafkaStreamingSource does. If the user knows the schema of the data the source should take the schema and be able to create a structured record through that schema from the byte[].

File: spark-plugins/src/test/java/co/cask/hydrator/plugin/spark/test/SparkPluginTest.java [hunk 13]
```diff
@@ -124,7 +126,8 @@ public static void setupTest() throws Exception {
     addPluginArtifact(NamespaceId.DEFAULT.artifact("spark-plugins", "1.0.0"), parents,
                       NaiveBayesTrainer.class, NaiveBayesClassifier.class,
                       KafkaStreamingSource.class, KafkaUtils.class, DefaultDecoder.class,
-                      TwitterStreamingSource.class);
+                      TwitterStreamingSource.class, InitialPositionInStream.class,
```
Comment (rohitsinha54) [inline spark-plugins/src/test/java/co/cask/hydrator/plugin/spark/test/SparkPluginTest.java:129]: This will just add the artifact where is the unit test which tests the KinesisStreamingSource ?

**Review by rohitsinha54 — CHANGES_REQUESTED** [review]

File: spark-plugins/pom.xml (hunk 7 above)
Comment (abhinavbansal246) [reply spark-plugins/pom.xml:27]: Kinesis spark streaming needs the newer version of spark streaming library, other spark streaming plugins dependent on this library are building and passing the unit tests.

File: spark-plugins/pom.xml (hunk 8 above)
Comment (abhinavbansal246) [reply spark-plugins/pom.xml:78]: kinesis streaming job requires this

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java (hunk 12 above)
Comment (russorat) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java]: @rohitsinha54 the data from kinesis is always returned as a json. JSON parsing isn't currently supported in the kafka streaming source (because it isn't supported by RecordFormats.java). So the reason we are always returning a string is because Kinesis is always returning JSON, and it doesn't make sense to give the user the option. When recordformats supports JSON, we can update the plugin.

File: spark-plugins/src/test/java/co/cask/hydrator/plugin/spark/test/SparkPluginTest.java (hunk 13 above)
Comment (abhinavbansal246) [reply spark-plugins/src/test/java/co/cask/hydrator/plugin/spark/test/SparkPluginTest.java:129]: I added unit tests for checking configurations

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java (hunk 11 above)
Comment (abhinavbansal246) [reply spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java:190]: made all the configuration fields final

File: spark-plugins/docs/KinesisSource-streamingsource.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+**appName:** The name of the kinesis application. The application name that is used to checkpoint the Kinesis sequence
+numbers in DynamoDB table
+
+**streamName:** The name of the Kinesis stream to the get the data from. The stream should be active.
+
+**duration:** The interval in milliseconds at which the Kinesis Client Library saves its position in the stream.
+
+**endpointUrl:** Valid Kinesis endpoints URL eg. kinesis.us-east-1.amazonaws.com
+
+**accessID:** The access Id provided by AWS required to access the kinesis streams. The Id can be stored in Cdap secure
```
Comment (rohitsinha54) [inline spark-plugins/docs/KinesisSource-streamingsource.md:23]: CDAP

File: spark-plugins/docs/KinesisSource-streamingsource.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+store and can be provided as macro configuration.
+
+**accessKey:** AWS access key secret having access to Kinesis streams. The key can be stored in Cdap secure store and
```
Comment (rohitsinha54) [inline spark-plugins/docs/KinesisSource-streamingsource.md:26]: CDAP

File: spark-plugins/docs/KinesisSource-streamingsource.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+can be provided as macro configuration.
+
+**initialPosition:** Initial position in the stream. Can be either TRIM_HORIZON or LATEST, Default position will be
```
Comment (rohitsinha54) [inline spark-plugins/docs/KinesisSource-streamingsource.md:29]: default initial  position will be

File: spark-plugins/pom.xml [hunk 17]
```diff
@@ -22,13 +22,11 @@
     <groupId>co.cask.hydrator</groupId>
     <version>1.5.0-SNAPSHOT</version>
   </parent>
-
```
Comment (rohitsinha54) [inline spark-plugins/pom.xml:25]: why remove the new lines ?

File: spark-plugins/pom.xml [hunk 18]
```diff
@@ -22,13 +22,11 @@
...
   <properties>
     <app.parents>system:cdap-data-pipeline[4.0.0-SNAPSHOT,4.1.0-SNAPSHOT),system:cdap-data-streams[4.0.0-SNAPSHOT,4.1.0-SNAPSHOT)</app.parents>
-    <spark.version>1.3.1</spark.version>
+    <spark.version>1.6.2</spark.version>
     <kafka.version>0.8.2.2</kafka.version>
   </properties>
-
```
Comment (rohitsinha54) [inline spark-plugins/pom.xml:31]: why remove the new lines ?

File: spark-plugins/pom.xml (hunk 8 above)
Comment (rohitsinha54) [reply spark-plugins/pom.xml:78]: Okay. Can you add exclusions of all the things which you don't need from if it contains dependencies

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+        return Schema.parseJson(schema);
+      } catch (IOException | NullPointerException e) {
+        throw new IllegalArgumentException("Invalid schema: " + e.getMessage());
+      }
+    }
+
+    public void validate() {
+      Schema messageSchema = parseSchema();
+      // if format is empty, there must be just a single message field of type bytes or nullable types.
+      if (Strings.isNullOrEmpty(format)) {
```
Comment (rohitsinha54) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java:272]: This whole code block till line 303 is duplicated in KafkaConfig. Will be better to refactor it

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+  public static class KinesisStreamConfig extends ReferencePluginConfig implements Serializable {
+
+    @Name("appName")
+    @Description(" The application name that will be used to checkpoint the Kinesis sequence numbers in DynamoDB table")
+    @Macro
+    private final String appName;
+
+    @Name("streamName")
+    @Description("The name of the Kinesis stream to the get the data from. The stream should be active")
+    @Macro
```
Comment (rohitsinha54) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java:197]: the readme doesn't say it can be macro. same for many others here

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+    private final String endpoint;
+
+    @Name("region")
+    @Description("AWS region specific to the stream")
+    @Macro
+    private final String region;
+
+    @Name("duration")
+    @Description("The interval (e.g., Duration(2000) = 2 seconds) at which the Kinesis Client Library saves its " +
```
Comment (rohitsinha54) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java:211]: so its in milliseconds ? Should be good to say that

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+        } catch (Exception e) {
+          throw new IllegalArgumentException(String.format(
+            "Unable to instantiate a message parser from format '%s' and message schema '%s': %s",
+            format, messageSchema, e.getMessage()), e);
+        }
+      }
+    }
+
+    private InitialPositionInStream getInitialPosition() {
+      if (initialPosition.toUpperCase().equals(InitialPositionInStream.TRIM_HORIZON.name())) {
```
Comment (rohitsinha54) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KinesisStreamingSource.java:307]: just to equals ignoring case

File: spark-plugins/src/test/java/co/cask/hydrator/plugin/spark/test/SparkPluginTest.java (hunk 13 above)
Comment (rohitsinha54) [reply spark-plugins/src/test/java/co/cask/hydrator/plugin/spark/test/SparkPluginTest.java:129]: How is the config test related to the source sink being added here ? 
There should be a unit test for the plugin ? Is it not possible to  use kinesis streaming source/sink in memory by starting a in memory kinesis stream or something ?

File: spark-plugins/widgets/KinesisSource-streamingsource.json [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+            "values": [
+              "LATEST",
+              "TRIM_HORIZON"
+            ],
+            "default": "LATEST"
+          }
+        },
+        {
+          "widget-type": "textbox",
+          "label": "AWS access key id",
```
Comment (rohitsinha54) [inline spark-plugins/widgets/KinesisSource-streamingsource.json:53]: Not sure how the UI works but if you are saying they will be stored in the secure store then these should not be textbox right ? There is some other widget for secure info

**Comment by russorat** [comment]
closing due to age
